### PR TITLE
[SystemZ][z/OS] Propagate IsText parameter to open text files as text

### PR DESCRIPTION
**Pull request by abhina-sree**
This patch adds an IsText parameter to the following functions openFileForRead, getBufferForFile, getBufferForFileImpl and determines whether a file is text by querying the file tag on z/OS. The default is set to OF_Text instead of OF_None, this change in value does not affect any other platforms other than z/OS.

### CODE REVIEW

**Review by zibi2 — APPROVED** [review]
LGTM

**Review by kadircet — COMMENTED** [review]
hi sorry for missing this during the review time, i believe this change is changing some of the core support library interfaces in a way that's not justified.

this is an interface implemented by quite a lot of both upstream and downstream clients, but this change doesn't describe what the semantics around `IsText` is. so none of these implementations have a way to sustainably understand and maintain this new semantic.

Moreover reading the commit description, I also cannot see how callers can figure out what to pass into this parameter. You were probably fine with this during the implementation as they don't need to care, but this will create an unreasonable maintenance burden for the contributors in the future as no one can reason about semantics here.

Even further, the way this functionality is implemented today, relies on doing system calls to figure out `IsText`ness. This is against the very nature of a VFS, the behavior you're implementing here will not compose well with rest of the contractual guarantees we have via the VFS abstractions (surely there're cases where we violate that already today, but that isn't a reason to add more).

---

So in the light of all of these, I believe this change should've gotten approvals from an LLVM code-owner before submission. In that regard, I can see that you've been trying for the past 2 weeks to get attention, unfortunately community can be slow to respond at times, I'd like to ask you to a little bit more patient, or even start an RFC in discourse to get more visibility (as PRs are quite easy to miss).

I was also making some recommendations in `RealFileSystem::openFileForRead` about how this change might be contained. I think it'd be a lot better if you can keep the interfaces clean and introduce the change only into the parts in which these semantics are meaningful.

---

Up until these questions are settled (and I am more than happy to follow up on the discussion on my behalf), may I ask you to revert this change? As it's going to create some confusion both in the upstream and downstream users the more it stays around.

**Comment by kadircet** [comment]
i've also just noticed https://github.com/llvm/llvm-project/commit/3b3accb598ec87a6a30b0e18ded06071030bb78f, which seem to be pushed without review and any tests, changing behavior more in a non-obvious way. Please not that logic you have in:
```
  Expected<file_t> FDOrErr = sys::fs::openNativeFileForRead(
      adjustPath(Name, Storage), IsText ? sys::fs::OF_Text : sys::fs::OF_None,
      &RealName);
```

is affecting all, not just zOS, and by fiddling with the value passed by default, you're in turn changing the flags used by all the platforms.

**Comment by abhina-sree** [comment]
> 3b3accb

Hi, sure I will revert the two changes so we can discuss further. The OF_Text flag only affects the z/OS platform behaviour, the OF_TextWithCRLF will affect z/OS and Windows. On other platforms, the flag doesn't show any difference in behaviour that I'm aware of

File: llvm/lib/Support/VirtualFileSystem.cpp [hunk 1]
```diff
@@ -323,10 +325,11 @@ ErrorOr<Status> RealFileSystem::status(const Twine &Path) {
 }
 
 ErrorOr<std::unique_ptr<File>>
-RealFileSystem::openFileForRead(const Twine &Name) {
+RealFileSystem::openFileForRead(const Twine &Name, bool IsText) {
   SmallString<256> RealName, Storage;
   Expected<file_t> FDOrErr = sys::fs::openNativeFileForRead(
-      adjustPath(Name, Storage), sys::fs::OF_None, &RealName);
+      adjustPath(Name, Storage), IsText ? sys::fs::OF_Text : sys::fs::OF_None,
```
[the remark this answers is not in the document]
Comment (abhina-sree) [reply llvm/lib/Support/VirtualFileSystem.cpp:331]: I will look into this and see if I hit any issues. I think if the file already exists, we can deduce it here, but in the case the file does not exist, we do not know whether the file is supposed to be text or binary without additional context

**Comment by kadircet** [comment]
thanks a lot for the swift response!

I can see how none of these implementations are not using those flags _today_, but we're changing the observable behavior for them as well, and if some of those implementations decides to give meaning to these flags, it might be impossible to keep it working. Especially because we're not clearly defining the semantics around how callers are going to figure those bits out.

So I am wondering:
- if we can make these changes in a very limited way, e.g. just in https://github.com/llvm/llvm-project/blob/main/llvm/lib/Support/Unix/Path.inc#L972-L1118. We already have lots of `__MVS__` specific regions in there. that's also the "right" layer if you want to perform syscalls.
- if we need to make those changes in the VFS interfaces indeed, can you provide the rationale explaining that need and what the new semantics mean both for implementors of the interface and the callers of it?

**Comment by abhina-sree** [comment]
> thanks a lot for the swift response!
> 
> I can see how none of these implementations are not using those flags _today_, but we're changing the observable behavior for them as well, and if some of those implementations decides to give meaning to these flags, it might be impossible to keep it working. Especially because we're not clearly defining the semantics around how callers are going to figure those bits out.
> 
> So I am wondering:
> 
> * if we can make these changes in a very limited way, e.g. just in https://github.com/llvm/llvm-project/blob/main/llvm/lib/Support/Unix/Path.inc#L972-L1118. We already have lots of `__MVS__` specific regions in there. that's also the "right" layer if you want to perform syscalls.
> * if we need to make those changes in the VFS interfaces indeed, can you provide the rationale explaining that need and what the new semantics mean both for implementors of the interface and the callers of it?

So when we call this function sys::fs::openNativeFileForRead, there is an assumption that we already know whether the file is text or binary based on the OpenFlags that are passed to it. This function RealFileSystem::openFileForRead also assumes the file is binary by passing the OF_None flag. I'm not sure if there is a better way to determine whether the file is text or binary in this function without querying the file tag using a z/OS syscall in this function. 

I think I can limit it to the following change to avoid adding additional parameters to all the other functions which I agree is a much cleaner and less invasive solution. Please let me know your thoughts, thanks!

```
diff --git a/llvm/lib/Support/VirtualFileSystem.cpp b/llvm/lib/Support/VirtualFileSystem.cpp
index 928c0b5a24ed..7153560754d1 100644
--- a/llvm/lib/Support/VirtualFileSystem.cpp
+++ b/llvm/lib/Support/VirtualFileSystem.cpp
@@ -22,6 +22,7 @@
 #include "llvm/ADT/Twine.h"
 #include "llvm/ADT/iterator_range.h"
 #include "llvm/Config/llvm-config.h"
+#include "llvm/Support/AutoConvert.h"
 #include "llvm/Support/Casting.h"
 #include "llvm/Support/Chrono.h"
 #include "llvm/Support/Compiler.h"
@@ -325,8 +326,16 @@ ErrorOr<Status> RealFileSystem::status(const Twine &Path) {
 ErrorOr<std::unique_ptr<File>>
 RealFileSystem::openFileForRead(const Twine &Name) {
   SmallString<256> RealName, Storage;
+  bool IsText = true;
+#ifdef __MVS__
+  // If the file is tagged with a text ccsid, it may require autoconversion.
+  llvm::ErrorOr<bool> IsFileText =
+      llvm::iszOSTextFile(Name.str().c_str());
+  if (IsFileText)
+    IsText = *IsFileText;
+#endif
   Expected<file_t> FDOrErr = sys::fs::openNativeFileForRead(
-      adjustPath(Name, Storage), sys::fs::OF_None, &RealName);
+      adjustPath(Name, Storage), IsText ? sys::fs::OF_Text : sys::fs::OF_None, &RealName);
   if (!FDOrErr)
     return errorToErrorCode(FDOrErr.takeError());
   return std::unique_ptr<File>(
```

**Comment by kadircet** [comment]
i think limiting this to `RealFileSystem::openFileForRead` LG, but can we make sure `IsText` defaults to `false` and we can only set it in `#ifdef __MVS__` block to make sure this is really a no-op for other platforms. e.g.:
```cppp
auto OpenFlags = sys::fs::OF_None;
#ifdef __MVS__
....
  OpenFlags |= sys::fs::OF_Text;
....
#endif
   Expected<file_t> FDOrErr = sys::fs::openNativeFileForRead(
      adjustPath(Name, Storage), OpenFlags, &RealName);
```

I'd still prefer doing this inside https://github.com/llvm/llvm-project/blob/main/llvm/lib/Support/Unix/Path.inc#L972-L1118 as we have other utility functions that create file-descriptors and all of them end up using these base functions. By having your logic at a layer above these, you risk having subtle discrepancies. But I don't know much about zOS or why this is needed, so I'll leave management of that risk to you.

**Comment by abhina-sree** [comment]
> OpenFlags |= sys::fs::OF_Text;

Thanks, I've made your suggested change and I opened a new PR here https://github.com/llvm/llvm-project/pull/109664. A short explanation for why distinguishing text and binary files on z/OS is important is because the native encoding is not UTF-8, so we need to know whether the file is text in order to convert to a UTF-8 encoding so that it is readable.

**Pull request merged**